[Feat] TourAPI 기능 구현

laigasus opened the pull request:
## 변경사항

TourAPI 서비스 제공

### Issue Link - #17

-----

## 체크리스트

리뷰 요청 전에 확인해야 할 사항들을 나열해주세요.

- [x] 내 코드를 스스로 검토했나요?
- [x] 핵심 기능에 대해 충분한 테스트를 수행했나요?
- [x] 분석이 필요한가요?

-----

## 참고
![image](https://github.com/Strong-Potato/TripVote-BE/assets/26517061/1aee93d0-6c5e-4e6e-a60d-60de38dc3fcf)

laigasus committed:
[Chore] 패키지 이동

Tourapi 에 수행될 클래스 분리
'+' place entity 처리를 위한 jpa 라이브러리

laigasus committed:
[Chore] 패키지 이동

Tourapi 에 수행될 클래스 분리
'-' place 관련 클래스 삭제

laigasus committed:
[Chore] 패키지 구성 설정 - openapi

`+` google
`+` tourapi -> 한국관광공사

laigasus committed:
[Feat] TourAPI 설정 환경 구성

`+` 인수값 모음 파일 추가

laigasus committed:
[Feat] TourAPI constant 추가

아래의 공용 정보가 담겨 있음:
- 아이템 카테고리(ContentTypeId)

laigasus committed:
[Feat] 엔티티 지원 유틸 추가 - enum

아이템 타입 검사 지원 도구

laigasus committed:
[Feat] 엔티티 지원 유틸 추가 - image

단일 칼럼 내 여러 이미지 데이터 저장 지원 도구

laigasus committed:
[Feat] 도시 검색 유틸 추가

행정구역 탐색 도우미

laigasus committed:
[Feat] Place 관련 엔티티 이동 및 보강

아래의 아이템을 제공함
- 숙소(Accommodation)
- 문화시설(Facility)
- 행사(Festival)
- 레포츠(Leports)
- 식당(Restaurant)
- 쇼핑(Shop)
- 관광 명소(Spot)

laigasus committed:
[Feat] TourAPI 반환 dto 폼 추가 - 공통

공통 응답 폼

laigasus committed:
[Feat] TourAPI 반환 dto 폼 추가 - same property

각 api 별 응답 폼 프로퍼티가 같은 것들만 모아두었음

laigasus committed:
[Feat] TourAPI 반환 dto 폼 추가 - diff property

각 api 별 응답 폼 프로퍼티가 다른 것들만 모아두었음

laigasus committed:
[Feat] TourAPI 반환 Mapper 추가

반환된 dto를 도메인 객체에 맞게 매핑

laigasus committed:
[Feat] 임시 오류 메시지 체계 구현

체게 정립 전 작성된 코드.
`@Deprecated` 해두었음

laigasus committed:
[Feat] Mapping -> 도메인 변환 도구 추가

laigasus committed:
[Feat] TourAPI 통신 도구 추가

엔드포인트 통신 목록:
- detail_intro1
- area_based_sync_list1
- detail_common1
- detail_image1
- search_keyword1

laigasus committed:
[Feat] TourAPI 통신 Service 추가

laigasus committed:
[Feat] 반환 객체 병합 도구 추가

TourAPI 서비스 반환 객체 병합 목적으로 제작

laigasus committed:
[Feat] Place 서비스 수행 테스트 작성

'+' PlaceService
'+' PlaceController

laigasus committed:
[Docs] Place 서비스 수행 http 테스트

Dr-KoKo commented on openapi/src/main/java/fc/be/domain/place/Place.java:
```diff
@@ -1,21 +1,32 @@
 
+@Getter
+@SuperBuilder
+@NoArgsConstructor(access = AccessLevel.PUBLIC)
```
1. 외부 api의 데이터를 바로 entity로 생성하게 되면 몇가지 위험이 있습니다.
   - 첫번째로, 데이터를 검증할 수가 없습니다. 저희 서비스는 장소 정보를 tourApi에 의존하기는 하지만 tourAPI의 데이터가 잘못되어 있을 가능성도 있을 것 같아요. 숙박기준인원을 요청했을 때 당연히 숫자값을 응답해줄 것이라고 예상했는데 '정보없음'과 같은 string으로 응답이 온다거나, 가격정보를 0로 응답해준다거나 하는 경우가 있을 것 같네요. 이 값들을 검증없이 저희 DB에 넣게 되면 비즈니스상 예상 못한 문제가 발생할 수 있을 것 같아요.
   - 두번째로, API의 응답형식이 변경되었을 때 entity가 수정되어야 합니다. 이렇게되면 place entity를 사용하는 저희 서비스의 모든 코드에 수정이 필요하게 되고 유지보수가 곤란해질 수 있을 것 같아요.

2. entity의 접근제어자: `Public`
   - TourAPIDomainConverter에서 xml 또는 json을 역직렬화하기 위해 생성자의 접근제어자를 public하게 열어두신 것 같습니다. 접근 제어자가 public하게 되면 (물론 팀원들은 믿지만) 코드를 짜다보면 나도 모르게 생성자를 호출할 가능성이 열리게 되는데요, 그러면 그 entity는 영속성 컨텍스트가 나도 모르게 DB에 넣어버리는 불상사가 생길 가능성이 생기는 것 같아요.
   - 특히 Place 엔티티는 단독으로 생성되어서는 안되는 `parent-table`입니다. `Place`가 `abstract class` -> `class` 로 변경되면서 단독으로 생성될 가능성이 열렸는데 그렇게되면 데이터의 정합성에도 영향이 갈 수 있을 것 같네요.


그럼에도 불구하고 도저히 설계상 곤란하다면 접근제어자는 `protected`로 유지하되, `reflection`을 이용해서 잠시 `accessibility`만 변경해서 사용하시는 방법이 좋을 것 같습니다!

laigasus replied:
@Dr-KoKo  [c0dbecf](https://github.com/Strong-Potato/TripVote-BE/pull/23/commits/c0dbecf9887afcfb6655afa16a46342412ac4a6c) 반영했습니다!

Dr-KoKo commented on openapi/build.gradle:
```diff
@@ -5,6 +5,31 @@ dependencies {
+    implementation 'com.fasterxml.jackson.datatype:jackson-datatype-jsr310:+'
+}
+
+def querydslDir = "build/generated/querydsl"
```
예전에는 이 설정이 있어야 Q클래스를 불러올 수 있었던 것으로 알고 있습니다. QueryDsl의 버전이 올라가면서 이제는 자동으로 해당 패키지에 Q클래스를 생성하게 바뀌었습니다.

laigasus replied:
@Dr-KoKo [b417930](https://github.com/Strong-Potato/TripVote-BE/pull/23/commits/b4179300e3d47b1bb7de26435b766ed7e29982dc) 에 반영했습니다!

Dr-KoKo commented on openapi/src/main/java/fc/be/domain/place/controller/PlaceController.java:
```diff
@@ -0,0 +1,29 @@
+
+@RestController
+@RequiredArgsConstructor
+public class PlaceController {
```
지금 저희 멀티모듈환경은 openapi를 외부 api와 우리 api를 연동하기 위한 유틸성(?) 모듈로 생각을 했어요. Controller가 openapi쪽에 있게 되면 사실상 place 도메인을 따로 빼서 MSA 구조로 가져가는 느낌이 드는 것 같네요. Controller는 app 모듈에 만들어 두는 것이 멀티모듈환경을 도입하게 된 원래 취지에 부합하는 것 같습니다.

Dr-KoKo replied:
마찬가지로 사실 `domain(entity)`도 `app`모듈에 위치해야한다는 생각이 들긴 하네요... tourapi의 응답을 entity로 변환하기 위해 `domain(entity)`를 `openapi`모듈에 두어야한다면 그냥 MSA 구조로 가져가는 게 맞는 것 같아요.

laigasus replied:
@Dr-KoKo app 모듈에서 구현 하려다가 openapi 까지 닿이게 되어서 gradle 에 impl project 했더니, 
순환참조 오류로 빌드가 안되더군요..

`domain` 모듈 만드기에는 당시 논의에서 만들지 말자고 했던 걸로 알아서 이리 되었네요... 
그래도 패키지 명은 동일하게 했습니다!

Dr-KoKo replied:
저도 좀 고민을 해봤는데 domain(`Place`)를 `openapi` 모듈에서 가져다 쓰는 것이 문제의 근원인 것 같네요. `openapi`모듈에서는 domain(`Place`)들 복붙 하여서 똑같은 형식으로 `dto`를 만들어서 사용하시고, `app`모듈에서는 entity 안에 fromEntity(Dto dto) 같은 static method를 만들어서 사용하는 것이 좋을 것 같네요. (어차피 `openapi`모듈을 import해올거니까 `dto`도 import 되겠져)

laigasus replied:
@Dr-KoKo 새 Task 에서 작업 수행하겠습니다!

Dr-KoKo replied:
#25 에 이슈 생성했습니다.

Dr-KoKo commented on openapi/src/main/java/fc/be/tourapi/constant/ContentTypeId.java:
```diff
@@ -0,0 +1,39 @@
+import static fc.be.tourapi.TourAPIProperties.*;
+
+@Getter
+@AllArgsConstructor
```
TourAPIProperties가 있는데 이 enum 클래스는 어떤 용도로 만든걸까요?

laigasus replied:
@Dr-KoKo place entity 에 EnumString 으로 사용하려고 만들었었습니다. 현재 Place 칼럼에도 반영이 되어 있구요.

어노테이션 중에서 상수로 필요로 하는 값이 있어서, TourAPIProperties 에 따로 만들었었습니다. 
그래서 사용 용도가 떨어진 것 같아요. 쓸 곳이 마땅하지 않으면 정리할 것 같습니다!

Dr-KoKo commented on openapi/src/main/java/fc/be/tourapi/dto/form/diff_property/detail_intro1/DetailIntro1Response.java:
```diff
@@ -0,0 +1,10 @@
+ * <b>아이템 단건 부가 정보(/detailIntro. 소개정보조회)</b>
+ * 상세소개 쉬는날, 개장기간 등 내역을 조회하는 기능
+ */
+public record DetailIntro1Response(Response<Item> response) {
```
DetailIntro'1'Response 에 '1'은 오타일까요?

laigasus replied:
> DetailIntro'1'Response 에 '1'은 오타일까요?

숫자 '1' 이거는 TourAPI 요청 이름 그대로 가져왔습니다
https://www.data.go.kr/data/15101578/openapi.do

Dr-KoKo replied:
아 이해했습니다

Dr-KoKo commented on openapi/src/main/java/fc/be/tourapi/dto/mapper/TourAPIMapper.java:
```diff
@@ -0,0 +1,18 @@
+ * @param <T> 장소 상위 도메인 객체. Place
+ * @param <I> TourAPI의 응답 객체
+ */
+public interface TourAPIMapper<T, I> {
```
`T`가 `Place`를 무조건 상속받아야하면 `extends Place`를 붙여두면 컴파일 시점에 에러가 잡힐 것 같아요.

Dr-KoKo commented on openapi/src/main/java/fc/be/tourapi/TourAPIProperties.java:
```diff
@@ -0,0 +1,49 @@
+    public static final String SHOP_STR = "38";//38:쇼핑
+    public static final String RESTAURANT_STR = "39";//39:음식점
+
+    private static final List<String> keyList = IntStream.rangeClosed(1, 5)
```
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

Dr-KoKo replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 개꿀

Dr-KoKo commented on openapi/src/main/java/fc/be/tourapi/TourAPIProperties.java:
```diff
@@ -0,0 +1,49 @@
+
+    private static int keyIndex = 0;
+
+    public static final String BASE_URL = "https://apis.data.go.kr/B551011/KorService1/";
```
멘토님 말씀하신 것처럼 설정값들은 yml로 빼는 것이 좋을 것 같네요. 아까 자현님이 cors관련 설정 말씀하실 때 썼던 `@ConfigurationProperties` 사용해서 가져다 쓰면 될 것 같아요.

Dr-KoKo commented on openapi/src/main/java/fc/be/tourapi/tools/TourAPIDomainConverter.java:
```diff
@@ -0,0 +1,99 @@
+            final Class<T> detailImageListClass
+    ) {
+        try {
+            T result = detailImageListClass.getDeclaredConstructor().newInstance();
+            if (result instanceof Place place) {
```
문제의 그 부분이네요.. @Nine-JH  의 도움이 필요할 것 같아요. 도와줘요 Nine-JH

Dr-KoKo commented on openapi/src/main/java/fc/be/tourapi/tools/ObjectMerger.java:
```diff
@@ -0,0 +1,37 @@
+            return objectMapper.convertValue(targetNode, clazz);
+        } catch (ClassCastException e) {
+            e.fillInStackTrace();
+            throw new ClassCastException("병합하는 객체들의 타입이 일치하지 않습니다.");
```
이 부분도 Java에서 제공하는 `exception`말고, 커스텀 exception으로 던져주면 좋을 것 같아요.

laigasus replied:
예외 들여오는대로 모두 반영하겠습니다!

Dr-KoKo commented on openapi/src/main/java/fc/be/tourapi/tools/ObjectMerger.java:
```diff
@@ -0,0 +1,37 @@
+            }
+            return objectMapper.convertValue(targetNode, clazz);
+        } catch (ClassCastException e) {
+            e.fillInStackTrace();
```
요거는 어떤 용도로 사용하신 걸까요?

laigasus replied:
tourapi에 응답하는 값들을 place 하위 아이템들에게 매핑되도록 설계했었습니다. 
그렇다 보니 매칭되지 않는 부분 값들이 있는데요(ex. 사진 조회시 상세정보는 못가져옴)
그래서 각 요청들에 대한 응답들(place1, place2, ..)을 합쳐서 사용하는 것이 어떨까 생각해서
이렇게 설계하게 되었습니다! 

처음에 각 객체별 dto 를 만들었었는데, 아이템수 * 요청 기능 수로 급격히 늘더군요...
그래서 저리 타협하기로 했었습니다

Dr-KoKo replied:
아하 이해했습니다. `e.fillInStackTrace()`는 빼셔도 될 것 같네요. 빼시고 추가로 ClassCastException 생성자에 넣어주세요.

Dr-KoKo reviewed (changes requested):
코맨트 남기기 미안할 정도로 너무 수고 많으셨어요 ㅠㅠ
충분히 휴식하시고 시간 날 때 코맨트 확인해주세요 :)

laigasus committed:
[Refactor] Place 관련 엔티티 생성자 접근 제어자 변경

제네릭 타입 변환 과정에서만 일시적으로 허용 후 잠그도록 설정

[ref](https://github.com/Strong-Potato/TripVote-BE/pull/23#discussion_r1442790373)

laigasus committed:
[Chore] gradle 설정 변경

`-` querydsl 부가 설정 제거

laigasus committed:
[Fix] Place 내 image 칼럼 수정

`final`

laigasus committed:
[Refactor] TourAPIMapper 수정

T extends Place

laigasus committed:
[Refactor] ObjectMerger 수정

`e`

laigasus committed:
[Build] gradle 라이브러리 추가 - main

`+` 설정파일을 가져올 수 있음

laigasus committed:
[Refactor] TourAPI 설정 값 조회 방식 변경

이제 상수값 조회가 아닌,
`application-openapi.yml` 에서 가져와 사용합니다

laigasus committed:
[Chore] 패키지 조정 - TourAPIProperties

TourAPIProperties 패키지 변경 반영

laigasus committed:
[Refactor] TourAPI 컴포넌트 사용체계 조정

TourAPIProperties로 인해 아래의 사용방식이 변경되었음
- TourAPICommunicator
- TourAPIService

laigasus committed:
[Chore] 패키지 조정 - TourAPI 키 변경 기능

키 변경 기능을 TourAPIProperties 에서 분리했음

laigasus committed:
[Fix] ObjectMerger 수정

불필요한 어노테이션 삭제

Dr-KoKo reviewed (approved):
수고하셨습니다. #25 에서 부탁드릴게요

happymink reviewed (approved):
👍
